students/TacSPx/lesson04/mailroompart2.py

```diff
@@ -0,0 +1,216 @@
+    while True:
+        donor_sort = {}
+        donor_sum = {}
+        donor_len = {}
+        donor_avg = {}
+        donor_merge = {}
+        # title
+        title = "*******************  XYZ Nonprofit Charity  ********************\n" \
+                "________________________________________________________________"
+        heading_name = ("| Donor Name:", "| Totals $:", "| #-of-Gifts:", "| Avg Gift $: ", "|")
+        header = f"{heading_name[0].upper():<20s} {heading_name[1].upper():<13s}" \
+                 f" {heading_name[2].upper():<6s} {heading_name[3].upper():<14s}|"
+        print(title)
+        print(header)
+        # define format row
+        row = "| {Name:<18s} | ${Total:<10.2f} | {NumbGifts:^11d} | ${AverageGifts:<10.2f} |".format
+        # revisions made below to add sorting by totals, step #1 set-up to sort, new list
+
+        # calcs in the dictionary
+        donor_sum = dict([(key, sum(values)) for key, values in donor_info.items()])
+        donor_len = dict([(key, len(values)) for key, values in donor_info.items()])
+        donor_avg = dict([(key, sum(values) / len(values)) for key, values in donor_info.items()])
+        # merge
+        for key in set(list(donor_sum.keys()) + list(donor_len.keys()) + list(donor_avg.keys())):
+            donor_merge.setdefault(key, []).append(donor_sum[key])
+            donor_merge.setdefault(key, []).append(donor_len[key])
+            donor_merge.setdefault(key, []).append(donor_avg[key])
+        # sort dictionary by totals
+        donor_sort = sorted(donor_merge.items(), key=itemgetter(1), reverse=True)
+        # report printing
+        for i in donor_sort:
+            print(row(Name=i[0], Total=i[1][0], NumbGifts=i[1][1], AverageGifts=i[1][2]))
+        print("________________________________________________________________")
+        input("Press Enter to Continue")
+        break
+
+
+###############################################
+#   #3 Send letters to all Donors/ download thank you files
+###############################################
+def all_thank_you():
+    ''' Prints thank you letters for all donors'''
+
+    print("This will create 'Thank You' letters for all donors")
+    while True:
+        choice = input("Are you sure?\nEnter 'y' or 'n': ")
+        if choice.lower() == 'y':
+            for i in donor_info:
+                donor_name = i.strip()
+                donor_amount = format(donor_info[i][-1], '.2f') # last donation received
+                #print(donor_name, donor_amount) # for testing
+                all_letters = letter(donor_name, donor_amount)
+                # creates the thank you letter in a text file
+                with open(donor_name + ".txt", "w") as f: # use donor name as file name
+                    f.write(all_letters)
+            input("\nCompleted creating Letters! Please check program directory for files \nPress Enter to Continue!")
+            break
+
+        elif choice.lower() != 'y':
+            main()
```

Comment:
if you want to return to main menu then have a `return` here - shouldn't call main() loop directly